src/processor/operator/persistent/reader_state.cpp

```diff
@@ -8,413 +8,115 @@ namespace kuzu {
-        auto numRowsToReturn = std::min(leftArrowArrays.getLeftNumRows(), DEFAULT_VECTOR_CAPACITY);
-        leftArrowArrays.appendToDataChunk(dataChunk, numRowsToReturn);
-        auto result = std::make_unique<ReaderMorsel>(currFileIdx, currBlockIdx, currRowIdx);
-        currRowIdx += numRowsToReturn;
-        return result;
-    }
-}
-
-void ReaderSharedState::readNextBlock(common::DataChunk* dataChunk) {
-    if (leftArrowArrays.getLeftNumRows() == 0) {
-        auto morsel = getMorselOfNextBlock();
-        if (morsel->fileIdx >= copyDescription->filePaths.size()) {
-            // No more blocks.
-            return;
-        }
-        if (morsel->fileIdx != readFuncData->fileIdx) {
-            readFuncData = initFunc(copyDescription->filePaths, morsel->fileIdx,
-                *copyDescription->csvReaderConfig, tableSchema);
-        }
-        readFunc(*readFuncData, morsel->blockIdx, dataChunk);
-        leftArrowArrays.appendFromDataChunk(dataChunk);
+template<ReaderSharedState::ReadMode READ_MODE>
+std::unique_ptr<ReaderMorsel> ReaderSharedState::getMorsel() {
+    std::unique_ptr<ReaderMorsel> morsel;
+    if constexpr (READ_MODE == ReaderSharedState::ReadMode::PARALLEL) {
+        mtx.lock();
     }
-}
-
-std::unique_ptr<ReaderMorsel> ReaderSharedState::getParallelMorsel() {
-    std::unique_lock xLck{mtx};
     while (true) {
-        auto morsel = getMorselOfNextBlock();
-        if (morsel->fileIdx >= copyDescription->filePaths.size()) {
+        if (currFileIdx >= copyDescription->filePaths.size()) {
             // No more blocks.
+            morsel = std::make_unique<ReaderMorsel>();
             break;
         }
-        assert(morsel->fileIdx < blockInfos.size() &&
-               morsel->blockIdx < blockInfos[morsel->fileIdx].numRowsPerBlock.size());
-        currRowIdx += blockInfos[morsel->fileIdx].numRowsPerBlock[morsel->blockIdx];
-        return morsel;
-    }
-    return std::make_unique<ReaderMorsel>();
-}
-
-std::unique_ptr<ReaderMorsel> ReaderSharedState::getMorselOfNextBlock() {
-    if (currFileIdx >= blockInfos.size()) {
-        return std::make_unique<ReaderMorsel>(currFileIdx, INVALID_BLOCK_IDX, INVALID_ROW_IDX);
-    }
-    auto numBlocksInFile = blockInfos[currFileIdx].numRowsPerBlock.size();
-    if (currBlockIdx >= numBlocksInFile) {
-        currFileIdx += copyDescription->fileType == CopyDescription::FileType::NPY ?
-                           copyDescription->filePaths.size() :
-                           1;
+        if (currBlockIdx < fileInfos[currFileIdx].numBlocks) {
+            morsel = std::make_unique<ReaderMorsel>(currFileIdx, currBlockIdx++);
+            break;
+        }
+        currFileIdx++;
         currBlockIdx = 0;
     }
-    return std::make_unique<ReaderMorsel>(currFileIdx, currBlockIdx++, currRowIdx);
+    if constexpr (READ_MODE == ReaderSharedState::ReadMode::PARALLEL) {
+        mtx.unlock();
+    }
+    return morsel;
 }
 
+template std::unique_ptr<ReaderMorsel>
```

Comment:
Should we move the declarations for template specialization to .h file?

Reply:
This is not declaration, but instantiation.